File: src/cmdbox/repositories/variable_repository.py
from typing import Sequence

from peewee import IntegrityError

from .base_repository import BaseRepository
from .errors import (
    ValidationError,
    NameConflictError,
    UnknownNameError,
    UnknownTagError,
    TagAttachError,
    TagDetachError,
)
from .validators import VariableValidator
from .results import TagAttachResult, TagDetachResult
from cmdbox.database import db
from cmdbox.models import Variable, Tag, VariableTag


class VariableRepository(BaseRepository[Variable]):
    model = Variable

    def __init__(self, validator: VariableValidator | None = None):
        self.validator = validator or VariableValidator()

    def create(self, name: str, value: str) -> Variable:
        """
        Validates and creates a new Variable object based on provided input parameters.

        Args:
            name (str): Unique identifier for the variable to be created.
            value (str): The value that will be subbed for the name when executing.

        Returns:
            Variable: The created Variable object.

        Raises:
            NameConflictError: If a variable with the provided name already exists
                and causes a unique constraint violation during creation.
            IntegrityError: If any database integrity issue occurs during the creation.
        """
        name = name.strip() if name else None
        self.validator.validate_create(name=name, value=value)
        try:
            return Variable.create(name=name, value=value)
        except IntegrityError as exc:
            if name is not None and self._is_unique_name_violation(exc):
                raise NameConflictError(name=name) from exc
            raise

    def get_by_name(self, name: str) -> Variable:
        """
        Retrieves a variable instance by its name.

        Converts the provided name to lowercase, searches for a variable with the
        matching name in the database, and retrieves it. This function raises an error
        if no variable with the specified name is found.

        Args:
            name (str): The name of the variable to retrieve.

        Returns:
            Variable: The variable object if found.

        Raises:
            UnknownNameError: If no variable with the specified name exists.
        """
        name = name.lower()
        var = Variable.get_or_none(Variable.name == name)
        if var is None:
            raise UnknownNameError(name=name)
        return var

    def update(self, var_name: str, **fields) -> Variable:
        """
        Updates an existing variable based on the provided name and fields.

        This method retrieves a variable by its name and updates its fields with the
        provided values. It validates the updates, ensuring field integrity and
        uniqueness. If the variable doesn't exist or an update violates constraints,
        appropriate actions are taken.

        Args:
            var_name: The name of the variable to update.
            **fields: Arbitrary keyword arguments representing the fields to update.
                Supported fields include 'name' and 'value'.

        Returns:
            Variable: The updated variable object if successful, or None if the
                variable is not found.

        Raises:
            ValidationError: If any of the provided fields are invalid or do not exist
                on a variable.
            NameConflictError: If the new name conflicts with an existing variable's
                name.
            IntegrityError: If there is a general integrity constraint violation during
                the update process.
        """
        var = self.get_by_name(var_name)
        if not fields:
            raise ValueError("No fields provided for update.")

        if "name" in fields and fields.get("name") is not None:
            fields["name"] = fields.get("name").strip()

        self.validator.validate_update(
            name=fields.get("name", var_name), value=fields.get("value", var.value)
        )

        try:
            for key, value in fields.items():
                if not hasattr(var, key):
                    raise ValidationError(f"Invalid field: {key}")
                if value is not None:
                    setattr(var, key, value)
            var.save()
            return var
        except IntegrityError as exc:
            name = fields.get("name", "")
            if name is not None and self._is_unique_name_violation(exc):
                raise NameConflictError(name=name) from exc
            raise

    def add_tags(self, name: str, tags: Sequence[str]) -> TagAttachResult:
        """
        Attach tags to a variable identified by its name.

        This function associates tags with a specified variable. If the tags already
        exist for the variable, they are added to an existing list. Otherwise, new tags
        are created and linked to the variable. If no tags are provided, it returns
        immediately with empty results. In case of a database integrity issue, an
        appropriate error is raised.

        Args:
            name (str): The name identifier of the variable to which the tags are to
                be attached.
            tags (Sequence[str]): A collection of tag names to be attached to the
                variable.

        Returns:
            TagAttachResult: An object containing lists of newly added tags and
                tags that already existed.

        Raises:
            TagAttachError: If there is an issue attaching the tags to the variable,
                typically due to database integrity constraints.
        """
        if not tags:
            return TagAttachResult(added=[], existing=[])
        tags_actual = self._get_tags_by_name(*tags)
        var = self.get_by_name(name)
        try:
            with db.atomic():
                added = []
                existing = []
                for tag in tags_actual:
                    var_tag, created = VariableTag.get_or_create(variable=var, tag=tag)
                    if created:
                        added.append(tag.name)
                    else:
                        existing.append(tag.name)
                return TagAttachResult(added=added, existing=existing)
        except UnknownTagError:
            raise
        except IntegrityError as exc:
            raise TagAttachError("Could not attach tags to variable.") from exc

    def remove_tags(self, name: str, tags: Sequence[str]) -> TagDetachResult:
        """
        Removes tags from a variable identified by the provided name. The method first validates
        the tags to ensure they exist in the database, then attempts to remove the associations
        between the variable and the respective tags. If a tag is not attached to the variable, it
        is recorded in the `not_attached` list, while successfully removed tags are recorded in
        the `removed` list. Any errors during detachment raise a `TagDetachError`.

        Args:
            name (str): The unique identifier or name of the variable from which the tags will
                be detached.
            tags (Sequence[str]): A list of tag names to be detached from the variable.

        Returns:
            TagDetachResult: Object that contains two lists:
                - `removed`: A list of successfully detached tags.
                - `not_attached`: A list of tags that were not associated with the variable.

        Raises:
            TagDetachError: If the detachment process encounters an issue, such as database
                integrity errors.
        """
        if not tags:
            return TagDetachResult(removed=[], not_attached=[])
        tags_actual = self._get_tags_by_name(*tags)
        variable = self.get_by_name(name)
        removed = []
        not_attached = []
        try:
            with db.atomic():
                for tag in tags_actual:
                    deleted = (
                        VariableTag.delete()
                        .where(
                            (VariableTag.variable == variable)
                            & (VariableTag.tag == tag)
                        )
                        .execute()
                    )
                    if deleted:
                        removed.append(tag.name)
                    else:
                        not_attached.append(tag.name)
        except IntegrityError as exc:
            raise TagDetachError("Could not detach tags from variable.") from exc
        return TagDetachResult(removed=removed, not_attached=not_attached)

    def list_all(
        self, order_by: str | Sequence[str] = "name", limit: int = 25
    ) -> list[Variable]:
        """
        Lists all variables from the database, optionally ordered by specified fields.

        Args:
            order_by (str | Sequence[str]): A string or sequence of strings indicating the field(s)
                by which the variables should be ordered. Defaults to "name".
            limit (int): The maximum number of variables to return. Defaults to 25.

        Returns:
            List[Variable]: A list of Variable objects retrieved from the database,
                sorted based on the specified ordering criteria.
        """
        ordering = self._resolve_ordering(order_by)
        return list(Variable.select().order_by(*ordering).limit(limit))

    def list_by_tag(
        self,
        tags: Sequence[str],
        order_by: str | Sequence[str] = "name",
        limit: int = 25,
    ) -> list[Variable]:
        """
        Fetches a list of variables filtered by specified tags and ordered by specific
        criteria.

        This method retrieves a list of `Variable` objects associated with the tags
        provided in the `tags` argument. The results are optionally ordered based on the
        `order_by` field and limited by the `limit` argument.

        Args:
            tags (Sequence[str]): A list of tag names to filter variables by.
            order_by (str | Sequence[str]): Criteria to order the resulting variable
                list. Defaults to "name".
            limit (int): The maximum number of variables to return. Defaults to 25.

        Returns:
            list[Variable]: A list of `Variable` objects matching the tags and ordered as
                requested.

        Raises:
            UnknownTagError: If one or more provided tags do not exist in the database.
        """
        tags_actual = self._get_tags_by_name(*tags)
        ordering = self._resolve_ordering(order_by)
        return list(
            Variable.select()
            .join(VariableTag)
            .where(VariableTag.tag << tags_actual)
            .order_by(*ordering)
            .distinct()
            .limit(limit)
        )

    def search(
        self, query: str, fields: str | Sequence[str] | None = "name"
    ) -> list[Variable]:
        """
        Searches for variables matching the given query across specified fields.

        This function performs a case-insensitive search for the query in the specified
        fields of the Variable model. It calculates a relevance score for each match
        based on the occurrence count and position of the query within the fields and
        sorts the results by relevance.
        Args:
            query (str): The search query to match in the specified fields.
            fields (str | Sequence[str] | None): The fields to search within. By
                default, searches within "name". Can be a single field name as a string
                or a sequence of field names.

        Returns:
            list[Variable]: A list of Variable objects matching the search query, sorted
                by relevance.

        Raises:
            ValueError: If any provided field does not exist on the Variable model.
        """
        return self._search(query, "name", fields)

    def delete(self, name: str) -> bool:
        """
        Deletes the variable with the specified name.

        Args:
            name (str): The name of the variable to delete.

        Returns:
            bool: True if the variable was deleted, False otherwise.
        """
        var = self.get_by_name(name)
        if not var:
            return False
        var.delete_instance()
        return True

    def _is_unique_variable_tag_violation(self, exc: IntegrityError) -> bool:
        msg = str(exc)
        return (
            "UNIQUE constraint failed" in msg
            and "variabletag.variable_id" in msg
            and "variabletag.tag_id" in msg
        )

File: internal/repository/variable_repository.go
package repository

import (
	"errors"
	"fmt"
	"strings"

	"github.com/PhantomLambSoft/CmdBox/internal/models"
	"github.com/PhantomLambSoft/CmdBox/internal/repository/validate"
	"gorm.io/gorm"
)

var variableOrderableColumns = map[string]bool{
	"id":         true,
	"name":       true,
	"value":      true,
	"created_at": true,
	"updated_at": true,
	"profile_id": true,
}

var defaultVariableSearchFields = []string{"name", "value"}

type VariableCreateConfig struct {
	Name      string
	Value     string
	ProfileID *uint
}

type VariableUpdateConfig struct {
	Name  *string
	Value *string
}

type VariableRepository interface {
	Create(input VariableCreateConfig) (*models.Variable, error)
	GetByName(name string, profileID *uint) (*models.Variable, error)
	GetByID(id uint, profileID *uint) (*models.Variable, error)
	Update(variable *models.Variable, input VariableUpdateConfig) (*models.Variable, error)
	Delete(variable *models.Variable) error

	ListAll(orderBy string, limit int, profileID *uint) ([]models.Variable, error)
	ListByTags(tag []models.Tag, orderBy string, limit int, profileID *uint) ([]models.Variable, error)
	Search(query string, fields []string, limit int, profileID *uint) ([]models.Variable, error)

	AddTags(variable *models.Variable, tags []models.Tag) (TagAttachResult, error)
	RemoveTags(variable *models.Variable, tags []models.Tag) (TagDetachResult, error)
}

type variableRepository struct {
	db          *gorm.DB
	profileRepo ProfileRepository
	validator   *validate.VariableValidator
}

func NewVariableRepository(db *gorm.DB, profileRepo ProfileRepository, validator *validate.VariableValidator) VariableRepository {
	if validator == nil {
		validator = validate.NewVariableValidator(nil)
	}
	return &variableRepository{db: db, profileRepo: profileRepo, validator: validator}
}

func (r *variableRepository) Create(input VariableCreateConfig) (*models.Variable, error) {
	name := strings.TrimSpace(input.Name)
	if err := r.validator.ValidateCreate(name, input.Value); err != nil {
		return nil, err
	}

	profileID, err := resolveVariableProfileID(input.ProfileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	variable := &models.Variable{
		Name:      name,
		Value:     input.Value,
		ProfileID: profileID,
	}

	if err := r.db.Create(variable).Error; err != nil {
		if isUniqueConstraintViolation(err, "variables", "name") {
			return nil, fmt.Errorf("%w: %q", ErrNameConflict, name)
		}
		return nil, fmt.Errorf("creating variable %q: %w", name, err)
	}
	return variable, nil
}

func (r *variableRepository) GetByName(name string, profileID *uint) (*models.Variable, error) {
	pid, err := resolveVariableProfileID(profileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	var variable models.Variable
	err = r.db.Where("name = ? AND profile_id = ?", name, pid).First(&variable).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, fmt.Errorf("%w: %q", ErrUnKnownName, err)
	}
	if err != nil {
		return nil, fmt.Errorf("looking up variable name: %q: %w", name, err)
	}
	return &variable, nil
}

func (r *variableRepository) GetByID(id uint, profileID *uint) (*models.Variable, error) {
	pid, err := resolveVariableProfileID(profileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	var variable models.Variable
	err = r.db.Where("id = ? AND profile_id = ?", id, pid).First(&variable).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, fmt.Errorf("%w: id %q", ErrUnknownVariable, err)
	}
	if err != nil {
		return nil, fmt.Errorf("looking up variable id: %d: %w", id, err)
	}
	return &variable, nil
}

func (r *variableRepository) Update(variable *models.Variable, input VariableUpdateConfig) (*models.Variable, error) {
	if variable == nil {
		return nil, ErrNoUpdateTarget
	}

	mergedName := variable.Name
	if input.Name != nil {
		mergedName = strings.TrimSpace(*input.Name)
	}
	mergedValue := variable.Value
	if input.Value != nil {
		mergedValue = strings.TrimSpace(*input.Value)
	}

	if err := r.validator.ValidateUpdate(&mergedName, &mergedValue); err != nil {
		return nil, err
	}

	if input.Name != nil {
		variable.Name = mergedName
	}
	if input.Value != nil {
		variable.Value = mergedValue
	}

	if err := r.db.Save(variable).Error; err != nil {
		if isUniqueConstraintViolation(err, "variables", "name") {
			return nil, fmt.Errorf("%w: %q", ErrNameConflict, mergedName)
		}
		return nil, fmt.Errorf("updating variable id: %d: %w", variable.ID, err)
	}
	return variable, nil
}

func (r *variableRepository) Delete(variable *models.Variable) error {
	if variable == nil {
		return nil
	}
	if err := r.db.Delete(variable).Error; err != nil {
		return fmt.Errorf("deleting variable id: %d: %w", variable.ID, err)
	}
	return nil
}

func (r *variableRepository) AddTags(variable *models.Variable, tags []models.Tag) (TagAttachResult, error) {
	if len(tags) == 0 {
		return TagAttachResult{}, nil
	}

	var added, existing []string
	err := r.db.Transaction(func(tx *gorm.DB) error {
		for _, tag := range tags {
			var link models.VariableTag
			err := tx.Where("variable_id = ? AND tag_id = ?", variable.ID, tag.ID).First(&link).Error
			switch {
			case errors.Is(err, gorm.ErrRecordNotFound):
				newLink := models.VariableTag{VariableID: variable.ID, TagID: tag.ID}
				if err := tx.Create(&newLink).Error; err != nil {
					return err
				}
				added = append(added, tag.Name)
			case err != nil:
				return err
			default:
				existing = append(existing, tag.Name)
			}
		}
		return nil
	})
	if err != nil {
		return TagAttachResult{}, fmt.Errorf("%w: %v", ErrTagAttachFailed, err)
	}
	return TagAttachResult{Added: added, Existing: existing}, nil
}

func (r *variableRepository) RemoveTags(variable *models.Variable, tags []models.Tag) (TagDetachResult, error) {
	if len(tags) == 0 {
		return TagDetachResult{}, nil
	}

	var removed, notAttached []string
	err := r.db.Transaction(func(tx *gorm.DB) error {
		for _, tag := range tags {
			result := tx.Where("variable_id = ? AND tag_id = ?", variable.ID, tag.ID).Delete(&models.VariableTag{})
			if result.Error != nil {
				return result.Error
			}
			if result.RowsAffected > 0 {
				removed = append(removed, tag.Name)
			} else {
				notAttached = append(notAttached, tag.Name)
			}
		}
		return nil
	})
	if err != nil {
		return TagDetachResult{}, fmt.Errorf("%w: %v", ErrTagDetachFailed, err)
	}
	return TagDetachResult{Removed: removed, NotAttached: notAttached}, nil
}

func (r *variableRepository) ListAll(orderBy string, limit int, profileID *uint) ([]models.Variable, error) {
	if orderBy == "" {
		orderBy = "name"
	}
	if limit <= 0 {
		limit = 25
	}
	orderClauses, err := resolveOrdering(orderBy, variableOrderableColumns)
	if err != nil {
		return nil, err
	}
	pid, err := resolveVariableProfileID(profileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	var variables []models.Variable
	q := r.db.Where("profile_id = ?", pid)
	for _, clause := range orderClauses {
		q = q.Order(clause)
	}
	if err := q.Limit(limit).Find(&variables).Error; err != nil {
		return nil, fmt.Errorf("listing variables: %w", err)
	}
	return variables, nil
}

func (r *variableRepository) ListByTags(tags []models.Tag, orderBy string, limit int, profileID *uint) ([]models.Variable, error) {
	if len(tags) == 0 {
		return nil, nil
	}
	if orderBy == "" {
		orderBy = "name"
	}
	if limit <= 0 {
		limit = 25
	}

	orderClauses, err := resolveOrdering(orderBy, variableOrderableColumns)
	if err != nil {
		return nil, err
	}
	pid, err := resolveVariableProfileID(profileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	tagIDs := make([]uint, len(tags))
	for i, t := range tags {
		tagIDs[i] = t.ID
	}

	var variables []models.Variable
	q := r.db.Distinct("variables.*").
		Joins("JOIN variable_tags ON variable_tags.variable_id = variables.id").
		Where("variable_tags.tag_id IN ? AND variables.profile_id = ?", tagIDs, pid)
	for _, clause := range orderClauses {
		q = q.Order(clause)
	}

	if err := q.Limit(limit).Find(&variables).Error; err != nil {
		return nil, fmt.Errorf("listing variables by tag: %w", err)
	}
	return variables, nil
}

func (r *variableRepository) Search(query string, fields []string, limit int, profileID *uint) ([]models.Variable, error) {
	if len(fields) == 0 {
		fields = defaultVariableSearchFields
	}
	if limit <= 0 {
		limit = 25
	}
	pid, err := resolveVariableProfileID(profileID, r.profileRepo)
	if err != nil {
		return nil, err
	}

	var variables []models.Variable
	searchInput := searchWithRelevanceInput{
		DB:                   r.db,
		Table:                "variables",
		Query:                query,
		Fields:               fields,
		AllowedColumns:       variableOrderableColumns,
		SecondaryOrderColumn: "name",
		Limit:                limit,
		ExtraWhere:           "profile_id = ?",
		ExtraArgs:            []interface{}{pid},
		Dest:                 &variables,
	}
	err = searchWithRelevance(searchInput)
	if err != nil {
		return nil, fmt.Errorf("searching variables: %w", err)
	}
	return variables, nil
}
